src/qibolab/_core/instruments/qcs.py

```diff
@@ -0,0 +1,343 @@
+                            ),
+                            envelope=pulse.envelope,
+                            amplitude=sweep_param_map.get("amplitude", pulse.amplitude),
+                            frequency=frequency,
+                            phase=sweep_param_map.get(
+                                "relative_phase", pulse.relative_phase
+                            ),
+                        )
+                        if pulse.envelope.kind == "drag":
+                            qcs_pulse = qcs_pulse.drag(coeff=pulse.envelope.beta)
+                    else:
+                        raise ValueError("Unrecognized pulse type", pulse.kind)
+
+                    pulses.append(qcs_pulse)
+
+                program.add_waveform(pulses, virtual_channel)
+
+            elif isinstance(channel, DcChannel):
+
+                pulses = []
+                sweep_param_map = sweeper_pulse_map.get(pulse.id, {})
+
+                for pulse in sequence.channel(channel_id):
+                    if pulse.kind == "delay":
+                        duration = (
+                            sweep_param_map.get(
+                                "duration", pulse.duration * NANOSECONDS
+                            ),
+                        )
+                    elif pulse.kind == "pulse":
+                        qcs_pulse = qcs.DCWaveform(
+                            duration=sweep_param_map.get(
+                                "duration", pulse.duration * NANOSECONDS
+                            ),
+                            envelope=generate_qcs_envelope(pulse.envelope),
+                            amplitude=sweep_param_map.get("amplitude", pulse.amplitude),
+                        )
+                    else:
+                        raise ValueError("Unrecognized pulse type", pulse.kind)
+
+                program.add_waveform(pulses, virtual_channel)
+        if hw_demod:
+            program = program.n_shots(num_shots)
+        return program
+
+    def play(
+        self,
+        configs: dict[str, Config],
+        sequences: list[PulseSequence],
+        options: ExecutionParameters,
+        sweepers: list[ParallelSweepers],
+    ) -> dict[int, Result]:
+
+        if options.relaxation_time is not None:
+            self.backend._init_time = int(options.relaxation_time)
+
+        ret: dict[PulseId, np.ndarray] = {}
+        for sequence in sequences:
+            results = self.backend.apply(
+                self.create_program(
+                    sequence.align_to_delays(), configs, sweepers, options.nshots
+                )
+            ).results
+            acquisitions = sequence.acquisitions
+            acquisition_map: dict[qcs.Channels, list[InputOps]] = {}
+
+            for channel_id, input_op in acquisitions:
+                channel = self.virtual_channel_map[channel_id]
+                if channel not in acquisition_map:
+                    acquisition_map[channel] = []
```

Comment:
```suggestion
            acquisition_map: dict[qcs.Channels, list[InputOps]] = defaultdict(list)

            for channel_id, input_op in acquisitions:
                channel = self.virtual_channel_map[channel_id]
```
https://docs.python.org/3/library/collections.html#collections.defaultdict

Reply:
Fixed in https://github.com/qiboteam/qibolab/pull/944/commits/e9d79f5f30e852071ea9173b9c89704776218112